rfcs/20250421-source-target-pairs.md

```diff
@@ -0,0 +1,136 @@
+  * `num_replicas` if `cross_replica` is used.
+  * `num_partitions` if `cross_partition` is used.
+* (C5) `channel_type` is defined as:
+  * `DEVICE_TO_HOST` if `is_host_transfer = true`,
+  * `DEVICE_TO_DEVICE` otherwise.
+
+#### Examples
+
+```mlir
+%result = "stablehlo.send"(%operand, %token) {
+  source_target_pairs = dense<[[0, 1], [1, 2]]> : tensor<2x2xi64>,
+  channel_handle = #stablehlo.channel_handle<handle = 1, type = 2>,
+  is_host_transfer = true
+} : (tensor<2x2xi64>, !stablehlo.token) -> !stablehlo.token
+```
+
+### recv
+
+#### Semantics
+
+Receives data from a channel with `channel_id` and produces `results`.
+
+If `is_host_transfer` is `true`, then the operation transfers data from the
+host. Otherwise, it transfers data from another device based on the values of
+`source_target_pairs`. This flag duplicates the information provided in
+`channel_type`, so in the future we are planning to only keep one of them
+([#666](https://github.com/openxla/stablehlo/issues/666)).
+
+`results` consist of payload values which come first and a token which comes
+last. In the future, we are planning to split the payload and the token into two
+separate outputs to improve clarity
+([#670](https://github.com/openxla/stablehlo/issues/670)).
+
+#### Inputs
+
+| Label | Name                  | Type                                            | Constraints |
+|-------|-----------------------|-------------------------------------------------|-------------|
+| (I1)  | `token`               | `token`                                         |             |
+| (I2)  | `source_target_pairs` | 2-dimensional tensor constant of type `si64`    | (C1-C4)     |
+| (I3)  | `channel_id`          | constant of type `si64`                         |             |
+| (I4)  | `channel_type`        | enum of `DEVICE_TO_DEVICE` and `DEVICE_TO_HOST` | (C5)        |
+| (I5)  | `is_host_transfer`    | constant of type `i1`                           | (C5)        |
+
+#### Outputs
+
+| Name      | Type                                                    | Constraints |
+|-----------|---------------------------------------------------------|-------------|
+| `results` | variadic number of tensors, quantized tensors or tokens | (C2-C4)     |
+
+#### Constraints
+
+* (C1) `dim(source_target_pairs, 1) = 2`.
+* (C2) `is_unique(source_target_pairs[:, 0])`.
+* (C3) `is_unique(source_target_pairs[:, 1])`.
+* (C4) `0 <= source_target_pairs < N`, where `N` is defined as:
+  * `num_replicas` if `cross_replica` is used.
+  * `num_partitions` if `cross_partition` is used.
+* (C5) `channel_type` is defined as:
+  * `DEVICE_TO_HOST` if `is_host_transfer = true`,
+  * `DEVICE_TO_DEVICE` otherwise.
+
+#### Examples
+
+```mlir
+%results0, %results1 = "stablehlo.recv"(%token) {
+  source_target_pairs = dense<[[0, 1], [1, 2]]> : tensor<2x2xi64>,
+  channel_handle = #stablehlo.channel_handle<handle = 1, type = 3>,
+  is_host_transfer = true
```

Comment:
Can we keep it `false` to make it clear that the `source_target_pairs` will have some effect in this particular example.

Reply:
suggestion applied